pkg/k8s/service_cache.go

```diff
@@ -420,6 +426,54 @@ func (s *ServiceCache) UniqueServiceFrontends() FrontendList {
 	return uniqueFrontends
 }
 
+// filterEndpoints filters local endpoints by using k8s service heuristics.
+// For now it only implements the topology aware hints.
+func (s *ServiceCache) filterEndpoints(localEndpoints *Endpoints, svc *Service) *Endpoints {
+	if !option.Config.EnableServiceTopology || svc == nil || !svc.TopologyAware {
+		return localEndpoints
+	}
+
+	if s.selfNodeZoneLabel == "" {
+		// The node doesn't have the zone label set, so we cannot filter endpoints
+		// by zone. Therefore, return all endpoints.
+		return localEndpoints
+	}
+
+	if svc.TrafficPolicy == loadbalancer.SVCTrafficPolicyLocal {
+		// According to https://kubernetes.io/docs/concepts/services-networking/topology-aware-hints/#constraints:
+		// """
+		// Topology Aware Hints are not used when either externalTrafficPolicy or
+		// internalTrafficPolicy is set to Local on a Service.
+		// """
+		return localEndpoints
+	}
+
+	filteredEndpoints := &Endpoints{Backends: map[string]*Backend{}}
+
+	for key, backend := range localEndpoints.Backends {
+		if len(backend.HintsForZones) == 0 {
+			break
```

Comment:
should this say `continue`?

Reply:
Discussed offline. Just for visibility pasting it here.

```
if len(backend.HintsForZones) == 0 {
			break
		}
```
This break is because of (https://kubernetes.io/docs/concepts/services-networking/topology-aware-hints/#safeguards):
> One or more endpoints does not have a zone hint: When this happens, the kube-proxy assumes that a transition from or to Topology Aware Hints is underway. Filtering endpoints for a Service in this state would be dangerous so the kube-proxy falls back to using all endpoints.

```
if len(filteredEndpoints.Backends) != 0 {
		return filteredEndpoints
	}
```
Otherwise, we would start dropping requests to the service on a node which zone doesn't match any hint. This also mentions that:
> A zone is not represented in hints: If the kube-proxy is unable to find at least one endpoint with a hint targeting the zone it is running in, it falls to using endpoints from all zones. This is most likely to happen as you add a new zone into your existing cluster

Reply:
:man_facepalming: above is wrong indeed. we should return all backends instead of breaking. fixed.